File: RxTango/python/examples/zip_attributes.py
"""Correlated read of two attributes, reporting their measured timestamp skew.

correlate_snapshot (built on rx.zip) fires both reads in parallel and emits
a single pair only when BOTH have returned.  If either fails, the error
propagates immediately — the pair is never half-processed.  The measured
skew between the two attributes' own DeviceAttribute.time values is
reported alongside the diff: two "simultaneous" reads can still be
measurably apart if the device serializes them internally.

Mirrors the Java `ZipAttributes` / `TangoTestCorrelate` example.

Usage:
    python zip_attributes.py [device]

    device  defaults to tango://localhost:10000/sys/tg_test/1
"""

import asyncio
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent / "src"))

from reactivex.scheduler.eventloop import AsyncIOScheduler

from rxtango import read_attribute_ts
from rxtango.correlate import correlate_snapshot


async def main() -> None:
    device = sys.argv[1] if len(sys.argv) > 1 else "tango://localhost:10000/sys/tg_test/1"

    loop      = asyncio.get_running_loop()
    scheduler = AsyncIOScheduler(loop)
    done      = asyncio.Event()

    print(f"Correlated snapshot from {device}  (single shot)\n")

    correlate_snapshot(
        read_attribute_ts(device, "double_scalar"),
        read_attribute_ts(device, "long_scalar"),
    ).subscribe(
        on_next=lambda c: print(
            f"  double_scalar = {c.values[0]:+.6f}\n"
            f"  long_scalar   = {c.values[1]}\n"
            f"  diff          = {c.values[0] - float(c.values[1]):+.6f}\n"
            f"  skew          = {c.skew:.4f}s"
        ),
        on_error=lambda e: (print(f"  ERROR: {e}", file=sys.stderr), done.set()),
        on_completed=done.set,
        scheduler=scheduler,
    )

    await done.wait()


if __name__ == "__main__":
    asyncio.run(main())

File: RxTango/cpp/examples/zip_attributes.cpp
/**
 * Atomic correlated reads — zip two attributes from the same device.
 *
 * Every tick fires two reads in parallel; the pair is only processed when
 * BOTH arrive.  If either fails the pair is dropped — no half-processed data.
 *
 * Mirrors Python's zip_attributes.py and Java's ZipAttributes.java / TangoTestCorrelate.java.
 *
 * Usage:
 *   ./zip_attributes [device] [attr1] [attr2] [interval-ms]
 *   defaults: sys/tg_test/1  double_scalar  double_scalar  500
 */

#include <chrono>
#include <csignal>
#include <iomanip>
#include <iostream>
#include <string>

#include <rxcpp/rx.hpp>
#include <rxtango/rxtango.hpp>

static volatile sig_atomic_t g_running = 1;

int main(int argc, char* argv[]) {
    const std::string device      = argc > 1 ? argv[1] : "tango://localhost:10000/sys/tg_test/1";
    const std::string attr1       = argc > 2 ? argv[2] : "double_scalar";
    const std::string attr2       = argc > 3 ? argv[3] : "double_scalar";
    const int         interval_ms = argc > 4 ? std::stoi(argv[4]) : 500;

    std::signal(SIGINT, [](int) { g_running = 0; });

    std::cout << "Correlated reads: " << attr1 << " × " << attr2
              << " every " << interval_ms << " ms  (Ctrl+C to stop)\n\n"
              << std::setw(20) << attr1 << "  " << std::setw(20) << attr2 << "\n"
              << std::string(44, '-') << "\n";

    // Mirrors Java: Flowable.interval().flatMapSingle(tick -> Single.zip(readA, readB, combiner))
    auto sub = rxcpp::observable<>::interval(std::chrono::milliseconds(interval_ms))
        .flat_map([device, attr1, attr2](long) {
            // zip two concurrent reads — pair only produced when BOTH complete
            return rxcpp::observable<>::zip(
                [](double a, double b) { return std::make_pair(a, b); },
                rxtango::read_attribute<double>(device, attr1),
                rxtango::read_attribute<double>(device, attr2)
            );
        })
        .subscribe(
            [](std::pair<double,double> p) {
                std::cout << std::setw(20) << p.first
                          << "  " << std::setw(20) << p.second << "\n";
            },
            [](std::exception_ptr e) {
                try { std::rethrow_exception(e); }
                catch (std::exception& ex) { std::cerr << "  ERROR: " << ex.what() << "\n"; }
            }
        );

    while (g_running) std::this_thread::sleep_for(std::chrono::milliseconds(100));
    sub.unsubscribe();
    return 0;
}
